pkg/flowaggregator/podcontroller/pod_controller.go

```diff
@@ -0,0 +1,150 @@
+	podListerSynced cache.InformerSynced
+	podIPStore      map[string]PodInfo
+	mutex           sync.Mutex
+}
+
+func NewPodController(
+	informerFactory informers.SharedInformerFactory,
+) *PodController {
+	podInformer := informerFactory.Core().V1().Pods()
+	controller := &PodController{
+		podInformer:     podInformer,
+		podLister:       podInformer.Lister(),
+		podListerSynced: podInformer.Informer().HasSynced,
+		podIPStore:      make(map[string]PodInfo),
+	}
+	podInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.addPod,
+			UpdateFunc: controller.updatePod,
+			DeleteFunc: controller.deletePod,
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (p *PodController) GetPodInfoByIP(podIP string) (PodInfo, bool) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	podInfo, exist := p.podIPStore[podIP]
+	if !exist {
+		klog.V(4).Infof("No pod is cached with the corresponding ip %s.", podIP)
+	}
+	return podInfo, exist
+}
+
+func (p *PodController) syncPodStore(action, podIP string, podInfo PodInfo) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if _, exists := p.podIPStore[podIP]; !exists && action == "DELETE" {
+		klog.V(4).Infof("Cannot find the old ip %s in store.", podIP)
+		return
+	}
+	if action == "UPDATE" {
+		p.podIPStore[podIP] = podInfo
+	} else {
+		delete(p.podIPStore, podIP)
+	}
+}
+
+func (p *PodController) addPod(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		klog.V(4).Info("Object does not have correct Pod interface")
+		return
+	}
+
+	klog.V(2).Infof("Processing Pod %s/%s ADD event", pod.Namespace, pod.Name)
+	if pod.Status.PodIP == "" {
+		// Can't add pod with IP as key when IP is unset.
+		klog.V(4).Infof("Pod %s/%s has not been assigned IP yet. Defer processing", pod.Namespace, pod.Name)
+		return
+	}
+	p.syncPodStore("UPDATE", pod.Status.PodIP, PodInfo{pod.Namespace, pod.Name, pod.Spec.NodeName})
+}
+
+func (p *PodController) updatePod(oldObj, curObj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	curPod := curObj.(*v1.Pod)
+	klog.V(2).Infof("Processing Pod %s/%s UPDATE event", curPod.Namespace, curPod.Name)
+	if curPod.Status.PodIP == oldPod.Status.PodIP {
+		return
+	}
+	p.syncPodStore("UPDATE", curPod.Status.PodIP, PodInfo{curPod.Namespace, curPod.Name, curPod.Spec.NodeName})
+}
+
+func (p *PodController) deletePod(old interface{}) {
+	pod, ok := old.(*v1.Pod)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting Pod, invalid type: %v", old)
+			return
+		}
+		pod, ok = tombstone.Obj.(*v1.Pod)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting Pod, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	klog.V(2).Infof("Processing Pod %s/%s DELETE event", pod.Namespace, pod.Name)
+	p.syncPodStore("DELETE", pod.Status.PodIP, PodInfo{pod.Namespace, pod.Name, pod.Spec.NodeName})
+}
+
+func (p *PodController) Run(stopCh <-chan struct{}) {
```

Comment:
Don't we need any watcher here?

Reply:
can you elaborate more on watcher you are referring to?

Reply:
If you look at the resource controller implementations in Antrea agent at pkg/agent/controller, they maintain a queue to track the events and a worker to process it rather than directly calling the function as part of the event handler. Watcher is probably a wrong term to describe it, where different kind of handling is being used in network policy controller.

https://github.com/antrea-io/antrea/blob/main/pkg/agent/controller/noderoute/node_route_controller.go#L100

Wanted to know if it is intentional to skip the queue maintenance here because we are confident that no event from go-client can be missed. Or do we need one here? Tagging @antoninbas @tnqn for more insights on this.

Reply:
There is no need for a workqueue if 1) there is no risk of transient errors that need to be handled with automatic retries, and 2) the amount of work you do in each handler is small. I think if you just need to update a map, you don't need a workqueue.

Reply:
If the method `GetPodInfoByIP` is the only thing you need, you don't have to create a pod controller to maintain a separate map.
Just pass `podInfomer` to `flowAggregator` and add a podIP indexer to the informer there.
See https://github.com/antrea-io/antrea/blob/e8ca54441b116c944accd3297eb7428234562b37/pkg/controller/traceflow/controller.go#L107

Reply:
Thanks for the info. Replaced pod controller with indexer.